### PR TITLE
TT-6335 backport TT-1849

### PR DESCRIPTION
**Pull request by sredxny**
## Description
Backporting the changes in https://github.com/TykTechnologies/tyk/pull/3490

## Related Issue
TT-6335

## Motivation and Context
provide smooth fix for customers

## How This Has Been Tested
added unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [ ] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

**Review by tbuchaillot — CHANGES_REQUESTED** [review]
@sredxny you are missing an important part in the `PurgeCache` part.

`
exp, _ := r.Store.GetExp(analyticsKey)
if exp == -1 {
	r.Store.SetExp(analyticsKey, int64(expireAfter))
            ...
`

the `if exp == -1` instead of `if exp <= 0` is key since we just want to set the expiration when it's not set.

**Comment by sredxny** [comment]
@tbuchaillot good catch!! done

File: gateway/redis_analytics_purger.go [hunk 1]
```diff
@@ -34,12 +34,17 @@ func (r RedisPurger) PurgeLoop(ctx context.Context) {
...
+		return
+	}
+
 	if expireAfter == 0 {
 		expireAfter = 60 // 1 minute
 	}
 
 	exp, _ := r.Store.GetExp(analyticsKeyName)
-	if exp <= 0 {
+	if exp <= -1 {
```
Comment (tbuchaillot) [inline gateway/redis_analytics_purger.go:47]: @sredxny this should be exp = -1. If TTL is -2, it means that the key doesn't exist so we don't want to set anything there.

**Review by tbuchaillot — CHANGES_REQUESTED** [review]

**Review by tbuchaillot — APPROVED** [review]
LGTM - please check standalone gw and mdcb scenarios before merge 👌

**Pull request merged**